pkg/kvstore/kvstore.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kvstore
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/vmware/vic/pkg/trace"
+)
+
+var (
+	ErrKeyNotFound = errors.New("key not found")
+)
+
+// This package implements a very basic key/value store.  It is up to the
+// caller to provision the namespace.
+
+type KeyValueStore struct {
+	b Backend
+
+	kv map[string][]byte
+
+	fileName string
+
+	l sync.Mutex
+}
+
+type Backend interface {
+	// Creates path and ovewrites whatever is there
+	Upload(ctx context.Context, r io.Reader, pth string) error
+
+	Download(ctx context.Context, pth string) (io.ReadCloser, error)
+
+	Mv(ctx context.Context, fromPath, toPath string) error
+}
+
+// Create a new KeyValueStore instance using the given Backend with the given
+// file.  If the file exists on the Backend, it is restored.
+func NewKeyValueStore(op trace.Operation, store Backend, fileName string) (*KeyValueStore, error) {
+	p := &KeyValueStore{
+		b:        store,
+		kv:       make(map[string][]byte),
+		fileName: fileName,
+	}
+
+	if err := p.restore(op); err != nil {
+		return nil, err
+	}
+
+	op.Infof("KeyValueStore(%s) restored %d keys", fileName, len(p.kv))
+
+	return p, nil
+}
+
+func (p *KeyValueStore) restore(op trace.Operation) error {
+	p.l.Lock()
+	defer p.l.Unlock()
+
+	rc, err := p.b.Download(op, p.fileName)
+	if err != nil {
+		// We need to check for 404 vs something else here.
+		op.Errorf("KeyValueStore(%s) ignoring error: %s", p.fileName, err)
+		return nil
+	}
+	defer rc.Close()
+
+	if err = json.NewDecoder(rc).Decode(&p.kv); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Add a key to the KeyValueStore with the given value.  The Add operation is
+// write-through consistent, which means it will sync the data to the backend
+// before updating the in-memory state and returning to the caller.
+func (p *KeyValueStore) Add(op trace.Operation, key string, value []byte) error {
+	p.l.Lock()
+	defer p.l.Unlock()
+
+	// get the old value in case we need to roll back
+	oldvalue, ok := p.kv[key]
+
+	if ok && bytes.Compare(oldvalue, value) == 0 {
+		// NOOP
+		return nil
+	}
+
+	p.kv[key] = value
```

Comment:
Doc-string says "it will sync the data to the backend before updating the in-memory state"

but what I see in the code that value is updated in memory fist and only then saved to back end.
Am I missing something? 

You could just store a value into map if "save" was successful, so no need to restore an old value.